docs/en/integrations/data-ingestion/data-formats/json.md

```diff
@@ -0,0 +1,603 @@
+└────────┴──────────────────────────────────────────────────────┘
+```
+
+Consider using [JSONAsObject](/docs/en/interfaces/schema-inference.md/#json-as-object) together with a new [JSON data type](/docs/en/guides/developer/working-with-json/json-semi-structured.md/#json-object-type) to store and process JSON in tables in a more efficient way. Note that JSONAsString works perfectly fine in cases we have JSON object-per-line formatted files (usually used with `JSONEachRow` format).
+
+## Data types detection when importing JSON data
+
+ClickHouse does some magic to guess the best types while importing JSON data. We can use a `DESCRIBE` clause to check which types were defined:
+
+```sql
+DESCRIBE TABLE file('list.json', JSONEachRow)
+```
+```response
+┌─name──┬─type─────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
+│ path  │ Nullable(String) │              │                    │         │                  │                │
+│ month │ Nullable(Date)   │              │                    │         │                  │                │
+│ hits  │ Nullable(Int64)  │              │                    │         │                  │                │
+└───────┴──────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘
+```
+
+This allows quickly creating tables from JSON files:
+
+```sql
+CREATE TABLE new_table
+ENGINE = MergeTree
+ORDER BY tuple() AS
+SELECT *
+FROM file('list.json', JSONEachRow)
+```
+
+Detected types will be used for this table:
+
+```sql
+DESCRIBE TABLE new_table
+```
+```response
+┌─name──┬─type─────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
+│ path  │ Nullable(String) │              │                    │         │                  │                │
+│ month │ Nullable(Date)   │              │                    │         │                  │                │
+│ hits  │ Nullable(Int64)  │              │                    │         │                  │                │
+└───────┴──────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘
+```
+
+
+## JSON objects with nested objects
+
+In cases we're dealing with [nested JSON objects](assets/list-nested.json), we can additionally define schema and use complex types ([Array](/docs/en/sql-reference/data-types/array.md/), [JSON](/docs/en/sql-reference/data-types/json.md/) or [Tuple](/docs/en/sql-reference/data-types/tuple.md/)) to load data:
+
+```sql
+SELECT *
+FROM file('list-nested.json', JSONEachRow, `page JSON, month Date, hits UInt32`)
+LIMIT 1
+```
+```response
+┌─page─────────────────────────────────────────────────────────────────────────┬──────month─┬─hits─┐
+│ {"owner_id":12,"path":"Akiba_Hebrew_Academy","title":"Akiba Hebrew Academy"} │ 2017-08-01 │  241 │
+└──────────────────────────────────────────────────────────────────────────────┴────────────┴──────┘
+```
+
```

Comment:
can we show the types inferred for the JSON field. Thanks for adding example +1

Reply:
This will just print `Object('json')`:
```
┌─name──┬─type───────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ page  │ Object('json') │              │                    │         │                  │                │
│ month │ Date           │              │                    │         │                  │                │
│ hits  │ UInt32         │              │                    │         │                  │                │
└───────┴────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘
```

It doesn't detect nested JSON automatically, so you have to specify it manually. Or am I missing something?

Reply:
describe_extend_object_types=1

Reply:
Only works with tables, not `file()`:
```sql
DESCRIBE TABLE file('list-nested.json', JSONEachRow, `page JSON, month Date, hits UInt32`)
SETTINGS describe_extend_object_types = 1

Query id: c85defb6-5b41-4b3f-8c26-87955e8d16fb

┌─name──┬─type───────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ page  │ Object('json') │              │                    │         │                  │                │
│ month │ Date           │              │                    │         │                  │                │
│ hits  │ UInt32         │              │                    │         │                  │                │
└───────┴────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

DESCRIBE TABLE test_json
SETTINGS describe_extend_object_types = 1

Query id: 5428c8d7-bc5f-4ac3-9b5a-bb7fbe612207

┌─name──┬─type─────────────────────────────────────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ page  │ Tuple(owner_id Int16, path String, title String) │              │                    │         │                  │                │
│ month │ Date                                             │              │                    │         │                  │                │
│ hits  │ UInt32                                           │              │                    │         │                  │                │
└───────┴──────────────────────────────────────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘
```

Do you think I should another section to mention that? Doesn't seem to be very practical, when we first have to load data into table to explore nested JSON structure. But once it works with `file()` - might be a powerful tool. Let me know your thoughts.

Reply:
Lets not block this merge and we'll address later.